### PR TITLE
Elizabeth/trie delete

### PR DESCRIPTION
**Pull request by noot**
## Changes
- implemented Delete from trie, can now delete nodes

## Tests:
```
go test ./trie/...
```

### CODE REVIEW

File: trie/trie.go [hunk 1]
```diff
@@ -187,3 +292,11 @@ func lenCommonPrefix(a, b []byte) int {
...
 	return length
 }
+
+// concat concatenates two byte arrays
+func concat(s1 []byte, s2 ...byte) []byte {
+	r := make([]byte, len(s1)+len(s2))
+	copy(r, s1)
+	copy(r[len(s1):], s2)
+	return r
+}
```
Comment (dutterbutter) [inline trie/trie.go]: `func concat` could be placed in a `common/common.go` file as it could be useful elsewhere and not only `trie.go`

Comment (noot) [reply trie/trie.go]: good move, will do

File: trie/trie_test.go [hunk 2]
```diff
@@ -93,3 +93,27 @@ func TestPutAndGet(t *testing.T) {
 		}
 	}
 }
+
+func TestDelete(t *testing.T) {
```
Comment (ansermino) [inline trie/trie_test.go]: I don't think these tests are sufficient. Try the following and see if you can improve the coverage:
```
go test -v ./... -coverprofile=coverage.out
go tool cover -html=coverage.out
```

Comment (noot) [reply trie/trie_test.go]: any ideas on how to improve coverage?

Comment (ansermino) [reply trie/trie_test.go]: You need to construct a larger tree and try deleting different parts from it. You are only ever deleting from an extension not a branch. You aren't testing invalid nodes, nor the case where `deleteFromExtension()` doesn't find the key immediately. Pretty sure the current test setup tests the exact same thing 100 times... You might consider constructing some more dynamic test tools, and checkout what the go-ethereum gang does.

Comment (noot) [reply trie/trie_test.go]: okay will do that. I think I am deleting from both branches and extensions but yeah I will try to make different testing tools

Comment (noot) [reply trie/trie_test.go]: @ansermino I have improved the testing, pls let me know what you think

**Comment by noot** [comment]
@ansermino @dutterbutter travis does not seem ok, any ideas?

**Comment by dutterbutter** [comment]
> @ansermino @dutterbutter travis does not seem ok, any ideas?

I do not think its Travis. Travis logs indicate one test is failing and on the build error, there seems to be a syntax issue.

**Comment by noot** [comment]
@ansermino ci passes

**Pull request merged**